File: torch_pythonvscpp/harness.cpp
// Timing harness, C++ / LibTorch side.
//
// Mirrors python/harness.py invariant-for-invariant:
//   * CUDA events for timing, never wall-clock around async calls.
//   * Warmup epochs / steps excluded from every statistic.
//   * Median + IQR for throughput; full percentile distribution for latency.
//   * Synchronise only at measurement boundaries.
//
// Emits JSON to stdout with the same schema as the Python side so a single
// aggregator (scripts/aggregate.py) consumes both. Config is read from the
// shared config/experiment.json.
//
// Build via CMake (see cpp/CMakeLists.txt). Requires the same libtorch that
// ships with the Python wheel used on the other side -- see README.

#include <torch/torch.h>

#include <algorithm>
#include <chrono>
#include <cmath>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

#include "resnet.h"
#include "json_mini.h"   // tiny header-only JSON reader/writer (see include/)

namespace {

struct Config {
    long seed = 42;
    std::string model = "resnet18";
    int num_classes = 10;
    bool cifar_stem = true;
    int batch_size = 128;
    double lr = 0.1, momentum = 0.9, weight_decay = 5e-4;
    std::string precision = "fp32";
    bool cudnn_benchmark = true;

    int warmup_steps = 50;
    int measured_steps = 200;
    int lat_warmup = 500;
    int lat_iters = 5000;
    int lat_batch = 1;
    int thr_batch = 512;
};

Config load_config(const std::string& path) {
    Config c;
    JsonValue j = json_parse_file(path);
    c.seed = j["seed"].as_int();
    c.model = j["model"].as_string();
    c.num_classes = j["num_classes"].as_int();
    c.cifar_stem = j["cifar_stem"].as_bool();
    c.batch_size = j["train"]["batch_size"].as_int();
    c.lr = j["train"]["lr"].as_double();
    c.momentum = j["train"]["momentum"].as_double();
    c.weight_decay = j["train"]["weight_decay"].as_double();
    c.precision = j["precision"].as_string();
    c.cudnn_benchmark = j["cudnn_benchmark"].as_bool();
    c.warmup_steps = j["harness"]["warmup_steps_per_epoch"].as_int();
    if (j["harness"].has("measured_steps"))
        c.measured_steps = j["harness"]["measured_steps"].as_int();
    c.lat_warmup = j["harness"]["inference_warmup_iters"].as_int();
    c.lat_iters = j["harness"]["inference_latency_iters"].as_int();
    c.lat_batch = j["harness"]["latency_batch_size"].as_int();
    c.thr_batch = j["harness"]["throughput_batch_size"].as_int();
    return c;
}

// CUDA-event timer; wall-clock fallback on CPU for local logic tests.
struct GpuTimer {
    bool cuda;
    at::cuda::CUDAEvent start_ev, end_ev;  // only used when cuda
    std::chrono::high_resolution_clock::time_point t0;

    explicit GpuTimer(bool is_cuda) : cuda(is_cuda) {}

    void start() {
        if (cuda) start_ev.record();
        else t0 = std::chrono::high_resolution_clock::now();
    }
    double stop_ms() {
        if (cuda) {
            end_ev.record();
            end_ev.synchronize();
            return start_ev.elapsed_time(end_ev);
        }
        auto t1 = std::chrono::high_resolution_clock::now();
        return std::chrono::duration<double, std::milli>(t1 - t0).count();
    }
};

double median(std::vector<double> v) {
    std::sort(v.begin(), v.end());
    size_t n = v.size();
    return n % 2 ? v[n / 2] : 0.5 * (v[n / 2 - 1] + v[n / 2]);
}

double percentile(std::vector<double> v, double p) {
    std::sort(v.begin(), v.end());
    size_t n = v.size();
    if (n == 1) return v[0];
    double rank = p / 100.0 * (n - 1);
    size_t lo = static_cast<size_t>(rank);
    double frac = rank - lo;
    size_t hi = std::min(lo + 1, n - 1);
    return v[lo] * (1 - frac) + v[hi] * frac;
}

void apply_precision(const Config& c) {
    at::globalContext().setBenchmarkCuDNN(c.cudnn_benchmark);
    if (c.precision == "fp32") {
        at::globalContext().setAllowTF32CuBLAS(false);
        at::globalContext().setAllowTF32CuDNN(false);
    } else if (c.precision == "tf32") {
        at::globalContext().setAllowTF32CuBLAS(true);
        at::globalContext().setAllowTF32CuDNN(true);
    }
    // bf16_amp handled at the autocast call site.
}

}  // namespace

int main(int argc, char** argv) {
    std::string config_path = "config/experiment.json";
    std::string arm, out_path;
    int repeat_index = 0;
    for (int i = 1; i < argc; ++i) {
        std::string a = argv[i];
        if (a == "--config" && i + 1 < argc) config_path = argv[++i];
        else if (a == "--arm" && i + 1 < argc) arm = argv[++i];
        else if (a == "--out" && i + 1 < argc) out_path = argv[++i];
        else if (a == "--repeat-index" && i + 1 < argc)
            repeat_index = std::stoi(argv[++i]);
    }

    Config c = load_config(config_path);
    torch::manual_seed(c.seed + repeat_index);
    torch::Device device(torch::cuda::is_available() ? torch::kCUDA
                                                     : torch::kCPU);
    bool is_cuda = device.is_cuda();
    apply_precision(c);
    bool amp = (c.precision == "bf16_amp");

    std::ostringstream js;
    js << "{\n";
    js << "  \"language\": \"cpp\",\n";
    js << "  \"arm\": \"" << arm << "\",\n";
    js << "  \"precision\": \"" << c.precision << "\",\n";
    js << "  \"model\": \"" << c.model << "\",\n";
    js << "  \"device\": \"" << (is_cuda ? "cuda" : "cpu") << "\",\n";
    js << "  \"repeat_index\": " << repeat_index << ",\n";

    auto autocast_guard = [&]() {
        // Return an optional autocast scope for bf16.
        return amp && is_cuda;
    };

    if (arm == "train_throughput") {
        ResNet model = build_model(c.model, c.num_classes, c.cifar_stem);
        model->to(device);
        model->train();
        torch::optim::SGD opt(model->parameters(),
            torch::optim::SGDOptions(c.lr).momentum(c.momentum)
                .weight_decay(c.weight_decay));

        int total = c.warmup_steps + c.measured_steps;
        std::vector<std::pair<torch::Tensor, torch::Tensor>> data;
        data.reserve(total);
        for (int i = 0; i < total; ++i) {
            auto x = torch::randn({c.batch_size, 3, 32, 32}).to(device);
            auto y = torch::randint(0, c.num_classes, {c.batch_size},
                                    torch::kLong).to(device);
            data.emplace_back(x, y);
        }

        std::vector<double> step_ms;
        GpuTimer timer(is_cuda);
        for (int i = 0; i < total; ++i) {
            timer.start();
            opt.zero_grad();
            torch::Tensor loss;
            if (autocast_guard()) {
                at::autocast::set_autocast_enabled(at::kCUDA, true);
                at::autocast::set_autocast_dtype(at::kCUDA, at::kBFloat16);
                auto out = model->forward(data[i].first);
                loss = torch::nn::functional::cross_entropy(out, data[i].second);
                at::autocast::clear_cache();
                at::autocast::set_autocast_enabled(at::kCUDA, false);
            } else {
                auto out = model->forward(data[i].first);
                loss = torch::nn::functional::cross_entropy(out, data[i].second);
            }
            loss.backward();
            opt.step();
            double ms = timer.stop_ms();
            if (i >= c.warmup_steps) step_ms.push_back(ms);
        }
        double med = median(step_ms);
        double q1 = percentile(step_ms, 25), q3 = percentile(step_ms, 75);
        double ips = c.batch_size / (med / 1e3);
        js << "  \"batch_size\": " << c.batch_size << ",\n";
        js << "  \"step_ms\": {\n";
        js << "    \"n\": " << step_ms.size() << ",\n";
        js << "    \"median\": " << med << ",\n";
        js << "    \"iqr\": " << (q3 - q1) << ",\n";
        js << "    \"min\": " << *std::min_element(step_ms.begin(), step_ms.end()) << ",\n";
        js << "    \"max\": " << *std::max_element(step_ms.begin(), step_ms.end()) << ",\n";
        js << "    \"images_per_sec_median\": " << ips << "\n";
        js << "  }\n";

    } else if (arm == "inference_latency") {
        ResNet model = build_model(c.model, c.num_classes, c.cifar_stem);
        model->to(device);
        model->eval();
        torch::NoGradGuard ng;
        auto x = torch::randn({c.lat_batch, 3, 32, 32}).to(device);

        std::vector<double> lat_ms;
        GpuTimer timer(is_cuda);
        int total = c.lat_warmup + c.lat_iters;
        for (int i = 0; i < total; ++i) {
            timer.start();
            if (autocast_guard()) {
                at::autocast::set_autocast_enabled(at::kCUDA, true);
                at::autocast::set_autocast_dtype(at::kCUDA, at::kBFloat16);
                auto out = model->forward(x);
                at::autocast::clear_cache();
                at::autocast::set_autocast_enabled(at::kCUDA, false);
            } else {
                auto out = model->forward(x);
            }
            double ms = timer.stop_ms();
            if (i >= c.lat_warmup) lat_ms.push_back(ms);
        }
        js << "  \"batch_size\": " << c.lat_batch << ",\n";
        js << "  \"iters\": " << lat_ms.size() << ",\n";
        js << "  \"latency_ms_median\": " << median(lat_ms) << ",\n";
        js << "  \"latency_ms_percentiles\": {\n";
        js << "    \"50\": " << percentile(lat_ms, 50) << ",\n";
        js << "    \"90\": " << percentile(lat_ms, 90) << ",\n";
        js << "    \"95\": " << percentile(lat_ms, 95) << ",\n";
        js << "    \"99\": " << percentile(lat_ms, 99) << ",\n";
        js << "    \"99.9\": " << percentile(lat_ms, 99.9) << "\n";
        js << "  }\n";

    } else if (arm == "inference_throughput") {
        int warm = 50;
        ResNet model = build_model(c.model, c.num_classes, c.cifar_stem);
        model->to(device);
        model->eval();
        torch::NoGradGuard ng;
        auto x = torch::randn({c.thr_batch, 3, 32, 32}).to(device);

        std::vector<double> batch_ms;
        GpuTimer timer(is_cuda);
        int total = warm + c.measured_steps;
        for (int i = 0; i < total; ++i) {
            timer.start();
            if (autocast_guard()) {
                at::autocast::set_autocast_enabled(at::kCUDA, true);
                at::autocast::set_autocast_dtype(at::kCUDA, at::kBFloat16);
                auto out = model->forward(x);
                at::autocast::clear_cache();
                at::autocast::set_autocast_enabled(at::kCUDA, false);
            } else {
                auto out = model->forward(x);
            }
            double ms = timer.stop_ms();
            if (i >= warm) batch_ms.push_back(ms);
        }
        double med = median(batch_ms);
        double q1 = percentile(batch_ms, 25), q3 = percentile(batch_ms, 75);
        js << "  \"batch_size\": " << c.thr_batch << ",\n";
        js << "  \"batch_ms\": {\n";
        js << "    \"n\": " << batch_ms.size() << ",\n";
        js << "    \"median\": " << med << ",\n";
        js << "    \"iqr\": " << (q3 - q1) << ",\n";
        js << "    \"images_per_sec_median\": "
           << (c.thr_batch / (med / 1e3)) << "\n";
        js << "  }\n";
    } else {
        std::cerr << "unknown arm: " << arm << "\n";
        return 2;
    }

    js << "}\n";
    std::cout << js.str();
    if (!out_path.empty()) {
        std::ofstream f(out_path);
        f << js.str();
    }
    return 0;
}

File: torch_pythonvscpp/harness.py
"""Timing harness, Python side.

Design invariants (mirrored exactly in cpp/src/harness.cpp):
  * Timing uses CUDA events, never wall-clock around async CUDA calls.
  * Warmup epochs and warmup steps are excluded from every statistic.
  * Every measured quantity is repeated and reported as median + IQR, never
    a single number and never the mean (throughput is right-skewed).
  * Inference latency is reported as a full percentile distribution
    (p50/p90/p95/p99/p99.9), because tail latency is the whole point of the
    low-latency argument.
  * The GPU is synchronised only at explicit measurement boundaries.

This file provides the measurement primitives and a CLI that runs the three
single-GPU arms: train throughput, inference latency, inference throughput.
It deliberately does NOT implement distributed training -- that is arm D and
lives separately so the single-GPU numbers stay uncontaminated.
"""
import argparse
import json
import statistics
import sys
import time

import torch

from model_path_shim_local import load_model_builder

build_model = load_model_builder()


# --------------------------------------------------------------------------
# measurement primitives
# --------------------------------------------------------------------------
class CudaTimer:
    """Elapsed GPU time in milliseconds between start() and stop(), using
    CUDA events. On CPU falls back to perf_counter (for local logic tests)."""

    def __init__(self, device):
        self.cuda = device.type == "cuda"
        if self.cuda:
            self._start = torch.cuda.Event(enable_timing=True)
            self._end = torch.cuda.Event(enable_timing=True)

    def __enter__(self):
        if self.cuda:
            self._start.record()
        else:
            self._t0 = time.perf_counter()
        return self

    def __exit__(self, *exc):
        if self.cuda:
            self._end.record()
            torch.cuda.synchronize()
            self.ms = self._start.elapsed_time(self._end)
        else:
            self.ms = (time.perf_counter() - self._t0) * 1e3


def percentiles(values, ps):
    s = sorted(values)
    out = {}
    n = len(s)
    for p in ps:
        if n == 1:
            out[p] = s[0]
            continue
        rank = p / 100.0 * (n - 1)
        lo = int(rank)
        frac = rank - lo
        hi = min(lo + 1, n - 1)
        out[p] = s[lo] * (1 - frac) + s[hi] * frac
    return out


def summarize(values):
    return {
        "n": len(values),
        "median": statistics.median(values),
        "iqr": (percentiles(values, [75])[75] - percentiles(values, [25])[25]),
        "min": min(values),
        "max": max(values),
    }


# --------------------------------------------------------------------------
# precision / backend setup
# --------------------------------------------------------------------------
def apply_precision(cfg):
    prec = cfg["precision"]
    torch.backends.cudnn.benchmark = cfg["cudnn_benchmark"]
    if prec == "fp32":
        torch.backends.cuda.matmul.allow_tf32 = False
        torch.backends.cudnn.allow_tf32 = False
        return None
    if prec == "tf32":
        torch.backends.cuda.matmul.allow_tf32 = True
        torch.backends.cudnn.allow_tf32 = True
        return None
    if prec == "bf16_amp":
        return torch.bfloat16
    raise ValueError(f"unknown precision {prec}")


# --------------------------------------------------------------------------
# arms
# --------------------------------------------------------------------------
def synthetic_loader(cfg, device, n_batches):
    """Synthetic data so the harness measures compute, not disk. The real
    dataset run uses the same loop with a torchvision loader; the synthetic
    path isolates the language overhead from the input pipeline (that is a
    separate arm, arm C)."""
    bs = cfg["train"]["batch_size"]
    g = torch.Generator(device="cpu").manual_seed(cfg["seed"])
    batches = []
    for _ in range(n_batches):
        x = torch.randn(bs, 3, 32, 32, generator=g)
        y = torch.randint(0, cfg["num_classes"], (bs,), generator=g)
        batches.append((x.to(device), y.to(device)))
    return batches


def arm_train_throughput(cfg, device):
    amp_dtype = apply_precision(cfg)
    h = cfg["harness"]
    model = build_model(cfg["model"], cfg["num_classes"],
                        cfg["cifar_stem"]).to(device).train()
    opt = torch.optim.SGD(model.parameters(),
                          lr=cfg["train"]["lr"],
                          momentum=cfg["train"]["momentum"],
                          weight_decay=cfg["train"]["weight_decay"])

    measured = h.get("measured_steps", 200)
    steps = h["warmup_steps_per_epoch"] + measured
    data = synthetic_loader(cfg, device, steps)
    bs = cfg["train"]["batch_size"]

    step_ms = []
    for i, (x, y) in enumerate(data):
        with CudaTimer(device) as t:
            opt.zero_grad(set_to_none=True)
            if amp_dtype is not None:
                with torch.autocast("cuda", dtype=amp_dtype):
                    loss = torch.nn.functional.cross_entropy(model(x), y)
            else:
                loss = torch.nn.functional.cross_entropy(model(x), y)
            loss.backward()
            opt.step()
        if i >= h["warmup_steps_per_epoch"]:
            step_ms.append(t.ms)

    s = summarize(step_ms)
    s["images_per_sec_median"] = bs / (s["median"] / 1e3)
    return {"arm": "train_throughput", "batch_size": bs,
            "precision": cfg["precision"], "step_ms": s}


def arm_inference_latency(cfg, device):
    amp_dtype = apply_precision(cfg)
    h = cfg["harness"]
    model = build_model(cfg["model"], cfg["num_classes"],
                        cfg["cifar_stem"]).to(device).eval()
    x = torch.randn(h["latency_batch_size"], 3, 32, 32, device=device)

    lat_ms = []
    with torch.no_grad():
        for i in range(h["inference_warmup_iters"] + h["inference_latency_iters"]):
            with CudaTimer(device) as t:
                if amp_dtype is not None:
                    with torch.autocast("cuda", dtype=amp_dtype):
                        _ = model(x)
                else:
                    _ = model(x)
            if i >= h["inference_warmup_iters"]:
                lat_ms.append(t.ms)

    pct = percentiles(lat_ms, [50, 90, 95, 99, 99.9])
    return {"arm": "inference_latency", "batch_size": h["latency_batch_size"],
            "precision": cfg["precision"], "iters": len(lat_ms),
            "latency_ms_percentiles": pct,
            "latency_ms_median": statistics.median(lat_ms)}


def arm_inference_throughput(cfg, device):
    amp_dtype = apply_precision(cfg)
    h = cfg["harness"]
    bs = h["throughput_batch_size"]
    model = build_model(cfg["model"], cfg["num_classes"],
                        cfg["cifar_stem"]).to(device).eval()
    x = torch.randn(bs, 3, 32, 32, device=device)

    measured = h.get("measured_steps", 200)
    warm = 50
    batch_ms = []
    with torch.no_grad():
        for i in range(warm + measured):
            with CudaTimer(device) as t:
                if amp_dtype is not None:
                    with torch.autocast("cuda", dtype=amp_dtype):
                        _ = model(x)
                else:
                    _ = model(x)
            if i >= warm:
                batch_ms.append(t.ms)
    s = summarize(batch_ms)
    s["images_per_sec_median"] = bs / (s["median"] / 1e3)
    return {"arm": "inference_throughput", "batch_size": bs,
            "precision": cfg["precision"], "batch_ms": s}


ARMS = {
    "train_throughput": arm_train_throughput,
    "inference_latency": arm_inference_latency,
    "inference_throughput": arm_inference_throughput,
}


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--config", default="../config/experiment.json")
    ap.add_argument("--arm", required=True, choices=list(ARMS))
    ap.add_argument("--out", default=None)
    ap.add_argument("--repeat-index", type=int, default=0,
                    help="which repeat this is; recorded in output for the "
                         "aggregator to compute cross-run medians")
    args = ap.parse_args()

    with open(args.config) as f:
        cfg = json.load(f)

    torch.manual_seed(cfg["seed"] + args.repeat_index)
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

    result = ARMS[args.arm](cfg, device)
    result["language"] = "python"
    result["torch_version"] = torch.__version__
    result["device"] = "cuda" if device.type == "cuda" else "cpu"
    result["repeat_index"] = args.repeat_index
    result["model"] = cfg["model"]

    text = json.dumps(result, indent=2)
    print(text)
    if args.out:
        with open(args.out, "w") as f:
            f.write(text)


if __name__ == "__main__":
    main()
